torch/distributed/distributed_c10d.py

```diff
@@ -1465,6 +1466,126 @@ def reduce_scatter(output,
+            list of tensors to scatter one per rank.
+            Either both `input` and `output` should be list of tensors or single tensor but not a mix
+        output_split_sizes: (list[Int], optional): Output split sizes for dim 0 when `output` is single tensor,
+            if specified None or empty, dim 0 of `output` tensor must divide equally by `world_size`. Ignored if 
+            `output` is a list of tensor
+        input_split_sizes: (list[Int], optional): Output split sizes for dim 0 when `input` is single tensor,
+            if specified None or empty, dim 0 of `input` tensor must divide equally by `world_size`. Ignored if 
+            `input` is a list of tensor
+        group (ProcessGroup, optional): The process group to work on
+        async_op (bool, optional): Whether this op should be an async op
+
+    Returns:
+        Async work handle, if async_op is set to True.
+        None, if not async_op or if not part of the group.
+
+    Examples::
+        >>> input = torch.arange(4) + rank * 4
+        >>> input
+        tensor([0, 1, 2, 3])     # Rank 0
+        tensor([4, 5, 6, 7])     # Rank 1
+        tensor([8, 9, 10, 11])   # Rank 2
+        tensor([12, 13, 14, 15]) # Rank 3
+
+        >>> output = torch.empty([4])
+        >>> dist.alltoall(output, input)
+        >>> output
+        tensor([0, 4, 8, 12])    # Rank 0
+        tensor([1, 5, 9, 13])    # Rank 1
+        tensor([2, 6, 10, 14])   # Rank 2
+        tensor([3, 7, 11, 15])   # Rank 3
+
+        >>> # Essentially, it is similar to following operation:
+        >>> scatter_list = list(input.chunk(world_size))
+        >>> gather_list  = list(output.chunk(world_size))
+        >>> for i in range(world_size):
+        >>>     dist.scatter(gather_list[i], scatter_list if i == rank else [], src = i)
+
+        >>> # Another example with uneven split
+        >>> input
+        tensor([0, 1, 2, 3, 4, 5])                                       # Rank 0
+        tensor([10, 11, 12, 13, 14, 15, 16, 17, 18])                     # Rank 1
+        tensor([20, 21, 22, 23, 24])                                     # Rank 2
+        tensor([30, 31, 32, 33, 34, 35, 36])                             # Rank 3
+        >>> input_splits
+        [2, 2, 1, 1]                                                     # Rank 0
+        [3, 2, 2, 2]                                                     # Rank 1
+        [2, 1, 1, 1]                                                     # Rank 2
+        [2, 2, 2, 1]                                                     # Rank 3
+        >>> output_splits
+        [2, 3, 2, 2]                                                     # Rank 0
+        [2, 2, 1, 2]                                                     # Rank 1
+        [1, 2, 1, 2]                                                     # Rank 2
+        [1, 2, 1, 1]                                                     # Rank 3
```

Comment:
Maybe worth mentioning that output_splits should be transpose of input_splits.

BTW, do we need this (uneven split) usecase for any applications?

We have no way of ensuring that user had specified correct splits (i.e. output_splits should be transpose of input_splits), so it can potentially lead to errors during MPI comm. Do you know if MPI detects this and what would happen?

Reply:
@agolynski  -  uneven splits where each rank can potentially send different amount of data to each rank is actually quite common. Which is why at least at the API level, we should make it as simple as possible to express (please see my comment above to compute the output splits from the input). MPI itself does not guarantee much safety in this regards and can crash the progress group silently - it's pretty implementation-dependent.

Reply:
@drdarshan Sounds good on uneven splits. I've left the comment there as well.

Reply:
```
        int sum = std::accumulate(split_sizes.begin(), split_sizes.end(), 0);
        if (sum != tensor.size(0)) {
          throw std::runtime_error("Split sizes doesn't match total dim 0 size");
        }
```
This is the check inside MPI code that throws error when split sizes do not match.